File: algorithm/104.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// needs to use fmax to reduce time overlimit
#define MAX(x, y) (x < y ? y : x)
int maxDepth(struct TreeNode *root)
{
    if (root == NULL)
    {
        return 0;
    }
    return MAX(maxDepth(root->left), maxDepth(root->right)) + 1;
}

File: go/goalgor/algor/104.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxHeight := 0
	curHeight := 0
	var dfs func(*TreeNode, *int)
	dfs = func(root *TreeNode, curHeight *int) {
		if root == nil {
			if maxHeight < *curHeight {
				maxHeight = *curHeight
			}
			return
		}
		*curHeight = *curHeight + 1
		dfs(root.Left, curHeight)
		dfs(root.Right, curHeight)
		*curHeight = *curHeight - 1
	}
	dfs(root, &curHeight)
	return maxHeight
}
